File: library/src/main/scala/com/dmanchester/playfop/internal/PlayFopImpl.scala
package com.dmanchester.playfop.internal

import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.io.StringReader

import scala.xml.Elem

import org.apache.fop.apps.FOUserAgent
import org.apache.fop.apps.Fop
import org.slf4j.LoggerFactory

import com.dmanchester.playfop.sapi.PlayFop

import javax.inject.Singleton
import javax.xml.transform.TransformerFactory
import javax.xml.transform.sax.SAXResult
import javax.xml.transform.stream.StreamSource
import play.twirl.api.Xml

@Singleton
class PlayFopImpl extends PlayFop {

  private val fopFactorySource = new FopFactorySource()

  private val logger = LoggerFactory.getLogger(this.getClass())

  def process[U](xslfo: Xml, outputFormat: String,
      autoDetectFontsForPDF: Boolean = false,
      foUserAgentBlock: (FOUserAgent => U) = {_: FOUserAgent => }): Array[Byte] = {

    logger.info("Rendering XSL-FO...")
    if (logger.isTraceEnabled()) {
      logger.trace(s"XSL-FO:\n$xslfo")
    }

    val output = new ByteArrayOutputStream()
    val fop = newFop(outputFormat, output, autoDetectFontsForPDF, foUserAgentBlock)

    val transformer = TransformerFactory.newInstance().newTransformer()

    val source = new StreamSource(new StringReader(xslfo.body))

    val result = new SAXResult(fop.getDefaultHandler())

    transformer.transform(source, result)

    val byteArray = output.toByteArray()

    logger.info(s"...XSL-FO rendered. ${byteArray.length} bytes produced.")

    byteArray
  }

  def newFop[U](outputFormat: String, output: OutputStream,
      autoDetectFontsForPDF: Boolean = false,
      foUserAgentBlock: (FOUserAgent => U) = {_: FOUserAgent => }): Fop = {

    val fopConfigXml: Option[Elem] = if (autoDetectFontsForPDF) {
      Some(
          <fop version="1.0">
            <renderers>
              <renderer mime="application/pdf">
                <fonts>
                  <auto-detect/>
                </fonts>
              </renderer>
            </renderers>
          </fop>
      )
    } else {
      None
    }

    val fopFactory = fopFactorySource.get(fopConfigXml)

    val foUserAgent = fopFactory.newFOUserAgent()
    foUserAgentBlock(foUserAgent)

    fopFactory.newFop(outputFormat, foUserAgent, output)
  }
}

File: library/src/main/java/com/dmanchester/playfop/internal_j/PlayFopImpl.java
package com.dmanchester.playfop.internal_j;

import java.io.OutputStream;

import javax.inject.Singleton;

import org.apache.fop.apps.FOUserAgent;
import org.apache.fop.apps.Fop;

import com.dmanchester.playfop.api_j.FOUserAgentBlock;
import com.dmanchester.playfop.api_j.PlayFop;
import com.dmanchester.playfop.api_j.ProcessOptions;

import play.twirl.api.Xml;
import scala.Function1;
import scala.runtime.AbstractFunction1;
import scala.runtime.BoxedUnit;

/**
 * The standard implementation of {@link PlayFop}.
 *
 * Instances of this class are thread-safe. They may be used across multiple
 * threads.
 */
@Singleton
public class PlayFopImpl implements PlayFop {

    private static final ProcessOptions DEFAULT_PROCESS_OPTIONS = new ProcessOptions.Builder().build();

    private com.dmanchester.playfop.api_s.PlayFop playFopScala = new com.dmanchester.playfop.internal_s.PlayFopImpl();

    @Override
    public byte[] process(Xml xslfo, String outputFormat) {

        return process(xslfo, outputFormat, DEFAULT_PROCESS_OPTIONS);
    }

    @Override
    public byte[] process(Xml xslfo, String outputFormat, ProcessOptions processOptions) {

        Function1<FOUserAgent, BoxedUnit> blockAsFunction = new BlockAsFunction(processOptions.getFoUserAgentBlock());

        return playFopScala.process(xslfo, outputFormat, processOptions.isAutoDetectFontsForPDF(), blockAsFunction);
    }

    @Override
    public Fop newFop(String outputFormat, OutputStream output) {

        return newFop(outputFormat, output, DEFAULT_PROCESS_OPTIONS);
    }

    @Override
    public Fop newFop(String outputFormat, OutputStream output, ProcessOptions processOptions) {

        Function1<FOUserAgent, BoxedUnit> blockAsFunction = new BlockAsFunction(processOptions.getFoUserAgentBlock());

        return playFopScala.newFop(outputFormat, output, processOptions.isAutoDetectFontsForPDF(), blockAsFunction);
    }

    private static class BlockAsFunction extends AbstractFunction1<FOUserAgent, BoxedUnit> {

        private FOUserAgentBlock block;

        public BlockAsFunction(FOUserAgentBlock block) {
            this.block = block;
        }

        @Override
        public BoxedUnit apply(FOUserAgent foUserAgent) {
            block.withFOUserAgent(foUserAgent);
            return BoxedUnit.UNIT;
        }
    }
}
